bothan-api/server-cli/src/commands/query.rs

```diff
@@ -0,0 +1,433 @@
+    connector: C,
+    ids: &[String],
+    timeout: Duration,
+) -> anyhow::Result<Vec<AssetInfo>>
+where
+    E1: StdError + Send + Sync + 'static,
+    E2: StdError + Send + Sync + 'static,
+    P: WebSocketAssetInfoProvider<SubscriptionError = E1, PollingError = E2>,
+    C: AssetInfoProviderConnector<Provider = P, Error = E1>,
+{
+    let mut provider = connector.connect().await?;
+    let deduped_ids = dedup_ids(ids);
+
+    provider.subscribe(&deduped_ids).await?;
+    query_websocket(&mut provider, deduped_ids, timeout).await
+}
+
+async fn query_websocket_in_chunks<C, P, E1, E2>(
+    connector: C,
+    ids: &[String],
+    max_subscription_per_connection: usize,
+    timeout: Duration,
+) -> anyhow::Result<Vec<AssetInfo>>
+where
+    E1: StdError + Send + Sync + 'static,
+    E2: StdError + Send + Sync + 'static,
+    P: WebSocketAssetInfoProvider<SubscriptionError = E1, PollingError = E2>,
+    C: AssetInfoProviderConnector<Provider = P, Error = E1>,
+{
+    let mut tasks = FuturesOrdered::new();
+    let deduped_ids = dedup_ids(ids);
+
+    for chunk in deduped_ids
+        .into_iter()
+        .chunks(max_subscription_per_connection)
+        .into_iter()
+    {
+        let chunk_ids = chunk.collect::<Vec<String>>();
+        let mut provider = connector.connect().await?;
+        WebSocketAssetInfoProvider::subscribe(&mut provider, &chunk_ids).await?;
+
+        tasks.push_back(async move { query_websocket(&mut provider, chunk_ids, timeout).await });
+    }
+
+    let results = tasks
+        .collect::<Vec<Result<Vec<AssetInfo>, anyhow::Error>>>()
+        .await;
+    let asset_infos = results
+        .into_iter()
+        .collect::<Result<Vec<Vec<AssetInfo>>, _>>()?
+        .into_iter()
+        .flatten()
+        .collect::<Vec<AssetInfo>>();
+
+    Ok(asset_infos)
+}
+
+async fn query_websocket<E1, E2, P>(
+    provider: &mut P,
+    ids: Vec<String>,
+    timeout_interval: Duration,
+) -> anyhow::Result<Vec<AssetInfo>>
+where
+    E1: StdError + Send + Sync + 'static,
+    E2: StdError + Send + Sync + 'static,
+    P: WebSocketAssetInfoProvider<SubscriptionError = E1, PollingError = E2>,
+{
+    let mut asset_infos: HashMap<String, AssetInfo> = HashMap::with_capacity(ids.len());
+
+    timeout(timeout_interval, async {
+        while asset_infos.len() < ids.len() {
+            if let Some(data) = provider.next().await {
+                if let Data::AssetInfo(infos) = data? {
+                    for info in infos {
+                        asset_infos.insert(info.id.clone(), info);
+                    }
+                }
+            } else {
+                return Err(anyhow!("stream closed unexpectedly"));
+            }
+        }
+
+        Ok(())
+    })
+    .await??;
+
+    ids.iter()
+        .map(|id| {
+            asset_infos
+                .remove(id)
+                .ok_or_else(|| anyhow!("missing asset info for id: {}", id))
+        })
+        .collect()
```

Comment:
given that the parsed ids here isn't the initial user input (its been deduped in the function before), i dont think ordering it with respect to the dedup is necessary. We can probably just collect the hashmap here